xlml/utils/gpu.py

```diff
@@ -304,49 +124,166 @@ def create_resource(
+    if accelerators:
+      instance.guest_accelerators = accelerators
+      instance.scheduling.on_host_maintenance = (
+          compute_v1.Scheduling.OnHostMaintenance.TERMINATE.name
+      )
+
+    if metadata:
+      instance.metadata = metadata
+
+    if service_account:
+      instance.service_accounts = [service_account]
+
+    if spot:
+      # Set the Spot VM setting
+      instance.scheduling.provisioning_model = (
+          compute_v1.Scheduling.ProvisioningModel.SPOT.name
+      )
+      instance.scheduling.instance_termination_action = instance_termination_action
+
+    if delete_protection:
+      # Set the delete protection bit
+      instance.deletion_protection = True
+
+    # Prepare the request to insert an instance.
+    request = compute_v1.InsertInstanceRequest()
+    request.zone = zone
+    request.project = project_id
+    request.instance_resource = instance
+
+    # Wait for the create operation to complete.
+    logging.info(f"Creating the {instance_name} instance in {zone}...")
+
+    operation = instance_client.insert(request=request)
+    return operation.name
+
+  @task.sensor(poke_interval=60, timeout=1800, mode="reschedule")
+  def wait_for_resource_creation(operation_name: airflow.XComArg):
+    # Retrives the delete opeartion to check the status.
+    client = compute_v1.ZoneOperationsClient()
+    request = compute_v1.GetZoneOperationRequest(
+        operation=operation_name,
+        project=project_id,
+        zone=zone,
+    )
+    operation = client.get(request=request)
+    status = operation.status.name
+    if status in ("RUNNING", "PENDING"):
+      logging.info(f"Resource create status: {status}, {operation.status_message}")
+      return False
+    else:
+      if operation.error:
+        logging.error(
+            f"Error during resource creation: [Code: {operation.http_error_status_code}]: {operation.http_error_message}",
+        )
+        raise operation.exception() or RuntimeError(operation.http_error_message)
+      if operation.warnings:
+        logging.warning(f"Warnings during resource creation:\n")
+        for warning in operation.warnings:
+          logging.warning(f" - {warning.code}: {warning.message}")
+      return True
 
-  ip_pattern = re.compile(r'network_i_p:\s+"([^"]+)"')
-  match = ip_pattern.search(str(instance))
+  @task
+  def get_ip_address(instance: str):
+    # It takes time to be able to use the ssh with the ip address even though the creation
+    # request is complete. We intentionally sleep for 60s to wait for the ip address to be
+    # accessible.
+    time.sleep(60)
+    instance_client = compute_v1.InstancesClient()
+    instance = instance_client.get(project=project_id, zone=zone, instance=instance)
+    if len(instance.network_interfaces) > 1:
+      logging.warning(f"GPU instance {gpu_name} has more than one network interface.")
+    return instance.network_interfaces[0].network_i_p
+
+  with TaskGroup(group_id="create_resource") as tg:
```

Comment:
Can you wrap `create_resource` in an `@task_group` instead?

Reply:
This is basically the task copy paste from https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/c1642e07dfba11953942d8c789ae9df6ffa8a002/xlml/utils/tpu.py#L182. I can update to @task_group. Let me update TPU/GPU both.

Reply:
I just updated the GPU here first to use task_group decorator. TPU update will be made in a separate PR in case anything goes wrong.